### PR TITLE
[CORE] Support compile and run on macos

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

This pr supports compile gluten on macos. Including 3 parts change:

- cpp, most change is to avoid using non-standard c++ api and make Apple clang compiler happy
- java, compatible with load lib, e.g., .so -> .dylib. Also bump protobuf, since the lower version plugin does not support macos
- bash,  compatible with macos and x86_64 or arm64 (m1)

Note, it does not support load lib from jar. It does not support compile hdfs or s3.

BTW, There are some small changes in oap/velox https://github.com/oap-project/velox/pull/388

## How was this patch tested?

manually test

image

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @zhztheplayer @zhejiangxiaomai @zhouyuan can you help review this pr ? also cc @Yohahaha  @CodingCat  I see you have some attempts before. thank you

File: cpp/core/jni/JniWrapper.cc [hunk 1]
```diff
@@ -872,10 +871,9 @@ JNIEXPORT jlong JNICALL Java_io_glutenproject_vectorized_ShuffleWriterJniWrapper
     std::string errorMessage = "Invalid shuffle writer id " + std::to_string(shuffleWriterId);
     throw gluten::GlutenException(errorMessage);
   }
-  jlong evictedSize;
-  gluten::arrowAssertOkOrThrow(
-      shuffleWriter->evictFixedSize(size, &evictedSize), "(shuffle) nativeEvict: evict failed");
-  return evictedSize;
+  int64_t* evictedSize = nullptr;
+  gluten::arrowAssertOkOrThrow(shuffleWriter->evictFixedSize(size, evictedSize), "(shuffle) nativeEvict: evict failed");
+  return reinterpret_cast<jlong>(evictedSize);
```
Comment (Yohahaha) [inline cpp/core/jni/JniWrapper.cc:876]: Does this safe?  seems before return jlong, after return int64_t pointer.

File: cpp/velox/CMakeLists.txt [hunk 2]
```diff
@@ -28,16 +28,21 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
...
+if (NOT ${CMAKE_SYSTEM_NAME} STREQUAL "Darwin")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-class-memaccess")
+endif()
 
 set(SYSTEM_LIB_PATH "/usr/lib" CACHE PATH "System Lib dir")
 set(SYSTEM_LIB64_PATH "/usr/lib64" CACHE PATH "System Lib64 dir")
 set(SYSTEM_LOCAL_LIB_PATH "/usr/local/lib" CACHE PATH "System Local Lib dir")
 set(SYSTEM_LOCAL_LIB64_PATH "/usr/local/lib64" CACHE PATH "System Local Lib64 dir")
-if (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)")
+if (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)|(arm64)")
```
Comment (Yohahaha) [inline cpp/velox/CMakeLists.txt:40]: Maybe a mistake? arm64 should in elseif branch.

File: cpp/velox/operators/plannodes/RowVectorStream.h [hunk 3]
```diff
@@ -76,7 +76,7 @@ class ValueStreamNode : public facebook::velox::core::PlanNode {
   }
 
  private:
-  void addDetails(std::stringstream& stream) const {};
+  void addDetails(std::stringstream& stream) const override{};
```
Comment (Yohahaha) [inline cpp/velox/operators/plannodes/RowVectorStream.h:79]: format problem, `override {}`

**Comment by Yohahaha** [comment]
Overall looks great, thanks!

File: cpp/core/jni/JniWrapper.cc (hunk 1 above)
Comment (ulysses-you) [reply cpp/core/jni/JniWrapper.cc]: this is a good point, fixed it

File: cpp/velox/CMakeLists.txt (hunk 2 above)
Comment (ulysses-you) [reply cpp/velox/CMakeLists.txt]: yeah, somehow I touched it

File: cpp/velox/operators/plannodes/RowVectorStream.h (hunk 3 above)
Comment (ulysses-you) [reply cpp/velox/operators/plannodes/RowVectorStream.h:79]: I like the style you point, but the style rule checker reject me. see the commit https://github.com/oap-project/gluten/pull/2691/commits/4b7592dcc4bf035426272ee5379d72c196c92385

**Comment by zhouyuan** [comment]
@ulysses-you great feature! thank you for improving this, could you please also add some notes for developer with new Mac here:
https://github.com/oap-project/gluten/blob/main/docs/developers/NewToGluten.md

Please note there are no CI running on Mac OS currently as it's difficult to get such resources in our lab. So this may rely on developers with M1/M2 devices to check :)

thanks, -yuan

**Comment by ulysses-you** [comment]
@zhouyuan I add the macOS at the support list. There should be no other requirements for developers.

File: cpp/velox/operators/plannodes/RowVectorStream.h [hunk 4]
```diff
@@ -59,7 +59,7 @@ class ValueStreamNode : public facebook::velox::core::PlanNode {
     return outputType_;
   }
 
-  const std::vector<facebook::velox::core::PlanNodePtr>& sources() const {
+  const std::vector<facebook::velox::core::PlanNodePtr>& sources() const override {
```
Comment (Yohahaha) [inline cpp/velox/operators/plannodes/RowVectorStream.h:62]: nit: curious why this format could pass format check?

File: cpp/velox/shuffle/VeloxShuffleReader.cc [hunk 5]
```diff
@@ -286,9 +286,9 @@ RowVectorPtr readRowVectorInternal(
     memory::MemoryPool* pool) {
   auto header = readColumnBuffer(batch, 0);
   uint32_t length;
-  mempcpy(&length, header->data(), sizeof(uint32_t));
+  memcpy(&length, header->data(), sizeof(uint32_t));
```
Comment (Yohahaha) [inline cpp/velox/shuffle/VeloxShuffleReader.cc:289]: Any particular reason for this?

File: ep/build-velox/src/build_velox.sh [hunk 6]
```diff
@@ -99,22 +117,42 @@ function compile {
   COMPILE_OPTION="$COMPILE_OPTION -DCMAKE_BUILD_TYPE=${BUILD_TYPE}"
   COMPILE_TYPE=$(if [[ "$BUILD_TYPE" == "debug" ]] || [[ "$BUILD_TYPE" == "Debug" ]]; then echo 'debug'; else echo 'release'; fi)
   echo "COMPILE_OPTION: "$COMPILE_OPTION
-  make $COMPILE_TYPE EXTRA_CMAKE_FLAGS="${COMPILE_OPTION}"
+
+  if [ $ARCH == 'x86_64' ]; then
+    make $COMPILE_TYPE EXTRA_CMAKE_FLAGS="${COMPILE_OPTION}"
+  elif [ $ARCH == 'arm64' ]; then
+    CPU_TARGET="arm64" make $COMPILE_TYPE EXTRA_CMAKE_FLAGS="${COMPILE_OPTION}"
```
Comment (Yohahaha) [inline ep/build-velox/src/build_velox.sh:124]: nit: not sure here may add a line breaker or use `cmdA && cmdB`?

File: ep/build-velox/src/build_velox.sh [hunk 7]
```diff
@@ -141,7 +179,20 @@ function check_commit {
   fi
 }
 
-function setup {
+function setup_macos {
+  if [ $ARCH == 'x86_64' ]; then
+    ./scripts/setup-macos.sh
+  elif [ $ARCH == 'arm64' ]; then
+    CPU_TARGET="arm64" ./scripts/setup-macos.sh
```
Comment (Yohahaha) [inline ep/build-velox/src/build_velox.sh:186]: same.

File: ep/build-velox/src/build_velox.sh (hunk 6 above)
Comment (Yohahaha) [inline ep/build-velox/src/build_velox.sh:124]: maybe export CPU_TARGET at top would make sense.

**Review by Yohahaha — COMMENTED** [review]
Some comments and nits, thanks!

File: ep/build-velox/src/build_velox.sh (hunk 6 above)
Comment (Yohahaha) [reply ep/build-velox/src/build_velox.sh:124]:
```
➜  ~ aa="bb" cat in.txt
hello
world
;;;;;%
➜  ~ echo $aa

➜  ~ aa="bb" && cat in.txt
hello
world
;;;;;%
➜  ~ echo $aa
bb
➜  ~
```

File: cpp/velox/shuffle/VeloxShuffleReader.cc (hunk 5 above)
Comment (ulysses-you) [reply cpp/velox/shuffle/VeloxShuffleReader.cc:289]: `mempcpy` is a gnu extension function, apple clang does not support that

File: ep/build-velox/src/build_velox.sh (hunk 6 above)
Comment (ulysses-you) [reply ep/build-velox/src/build_velox.sh:124]: I follow the meta/velox docs. see https://github.com/facebookincubator/velox/#setting-up-on-macos

File: cpp/velox/operators/plannodes/RowVectorStream.h (hunk 4 above)
Comment (ulysses-you) [reply cpp/velox/operators/plannodes/RowVectorStream.h:62]: not sure, I just apply `clang-format -i`

**Comment by ulysses-you** [comment]
@zhouyuan @Yohahaha  I'm going to merge this if no any more comments.

**Comment by zhouyuan** [comment]
@ulysses-you can you please do a rebase? The changes on Velox part is merged

**Review by zhouyuan — APPROVED** [review]
passed all Velox checks 👍 
also ping @zzcclp

**Pull request merged**